File: ios/LaneShadow/Views/Molecules/ErrorToast.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Error Toast Component

/**
 * Error Toast molecule component
 *
 * Displays error messages with an icon, title, description, and optional close button.
 * Following React Native component from react-native/components/toasts/error-toast.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Background: `theme.colors.danger.default`
 *   - Text/Icons: `theme.colors.onPrimary.default`
 * - Layout:
 *   - Corner radius: `theme.radius.lg`
 *   - Padding: `theme.space.md` (internal padding)
 *   - Gap: `theme.space.xs` (vertical spacing between elements)
 *   - Margin: `theme.space.sm` (horizontal margins), `theme.space.md` (vertical margins from safe area top)
 * - Shadow: Color black with 4pt offset, 0.15 opacity, 8pt radius
 * - Typography:
 *   - Title: `theme.type.title.sm`
 *   - Description: `theme.type.body.sm`
 *
 * ## Parameters
 * - title: Error title text
 * - description: Error description text
 * - showCloseButton: Whether to show close button (default: true)
 * - onClose: Optional callback when close button is tapped
 */
public struct LSErrorToast: View {
    @Environment(\.theme) private var theme

    private let title: String
    private let description: String
    private let showCloseButton: Bool
    private let onClose: (() -> Void)?

    public init(
        title: String,
        description: String,
        showCloseButton: Bool = true,
        onClose: (() -> Void)? = nil
    ) {
        self.title = title
        self.description = description
        self.showCloseButton = showCloseButton
        self.onClose = onClose
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: theme.space.xs) {
            // Header: icon + title + optional close button
            HStack(alignment: .top, spacing: theme.space.sm) {
                // Icon and title
                HStack(alignment: .top, spacing: 8) {
                    // Error icon
                    Image(systemName: "xmark.circle")
                        .font(.system(size: 20))
                        .foregroundStyle(theme.colors.onPrimary.default)

                    // Title
                    Text(title)
                        .font(theme.type.title.sm.font)
                        .foregroundStyle(theme.colors.onPrimary.default)
                }

                Spacer()

                // Optional close button
                if showCloseButton {
                    Button(action: {
                        onClose?()
                    }) {
                        Image(systemName: "xmark.circle")
                            .font(.system(size: 20))
                            .foregroundStyle(theme.colors.onPrimary.default)
                    }
                    .buttonStyle(PlainButtonStyle())
                    .accessibilityLabel("Close error toast")
                }
            }

            // Description
            Text(description)
                .font(theme.type.body.sm.font)
                .foregroundStyle(theme.colors.onPrimary.default)
        }
        .padding(theme.space.md)
        .background(theme.colors.danger.default)
        .clipShape(RoundedRectangle(cornerRadius: theme.radius.lg))
        .shadow(
            color: theme.elevation.level2.shadowColor,
            radius: theme.elevation.level2.radius,
            x: theme.elevation.level2.offsetX,
            y: theme.elevation.level2.offsetY
        )
        .accessibilityElement(children: .combine)
        .accessibilityLabel("Error: \(title). \(description)")
    }
}

// MARK: - Preview

#Preview("ErrorToast - With Close Button") {
    VStack(spacing: 16) {
        LSErrorToast(
            title: "Error",
            description: "Something went wrong. Please try again."
        )

        LSErrorToast(
            title: "Network Error",
            description: "Unable to connect to the server. Check your internet connection."
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("ErrorToast - Without Close Button") {
    VStack(spacing: 16) {
        LSErrorToast(
            title: "Error",
            description: "Something went wrong.",
            showCloseButton: false
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("ErrorToast - With Callback") {
    LSErrorToast(
        title: "Error",
        description: "This toast has a close callback.",
        onClose: {
            print("Close button tapped")
        }
    )
    .padding()
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/ErrorToast.kt
package com.laneshadow.ui.components.molecules

import com.laneshadow.ui.atoms.Glyphs

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * ErrorToast molecule component
 *
 * Red-themed toast for error notifications.
 * Displays error title, description, and optional close button.
 * Following React Native wrapper from react-native/components/toasts/error-toast.tsx
 *
 * @param title Error title text
 * @param description Error description text
 * @param showCloseButton Whether to show close button (default: true)
 * @param onClose Callback when close button is clicked
 * @param modifier Modifier for the toast container
 */
@Composable
fun ErrorToast(
    title: String,
    description: String,
    showCloseButton: Boolean = true,
    onClose: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    // Background color from danger token
    val backgroundColor = theme.colors.danger.default

    // Text/icon color from onPrimary token
    val contentColor = theme.colors.onPrimary.default

    // Build semantics with accessibility description
    val toastModifier = modifier.semantics {
        this.contentDescription = "Error: $title. $description"
    }

    Surface(
        modifier = toastModifier,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.lg),
        color = backgroundColor,
        shadowElevation = theme.elevation.light.level4,
    ) {
        Column(
            modifier = Modifier.padding(theme.space.md),
            verticalArrangement = Arrangement.spacedBy(theme.space.xs),
        ) {
            // Header row: icon + title + close button
            Row(
                modifier = Modifier,
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // Icon + title row
                Row(
                    modifier = Modifier.weight(1f),
                    horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    // Error icon
                    Icon(
                        imageVector = Glyphs.Default.Warning,
                        contentDescription = "Error icon",
                        tint = contentColor,
                        modifier = Modifier.size(20.dp),
                    )

                    // Title text
                    Text(
                        text = title,
                        style = theme.type.title.sm,
                        color = contentColor,
                    )
                }

                // Optional close button
                if (showCloseButton) {
                    IconButton(
                        onClick = { onClose?.invoke() },
                        modifier = Modifier.semantics {
                            this.contentDescription = "Close error toast"
                        },
                    ) {
                        Icon(
                            imageVector = Glyphs.Default.Close,
                            contentDescription = "Close",
                            tint = contentColor,
                            modifier = Modifier.size(20.dp),
                        )
                    }
                }
            }

            // Description text
            Text(
                text = description,
                style = theme.type.body.sm,
                color = contentColor,
            )
        }
    }
}
